triplea-game/triplea
# Move Substance dependencies to game-headed project

## Overview

Moves the Substance dependencies to the `game-headed` project in order to reduce the size of the `game-headless` and `lobby` archive artifacts even further:

Artifact | Old Size (MiB) | New Size (MiB) | Change
:-- | :-- | :-- | :--
`game-headless` | 18.5 | 17.1 | -8%
`lobby` | 12.0 | 10.6 | -12%

## Functional Changes

None.

## Refactoring Changes

* Broke a circular dependency between the initialization of the `ClientSetting` and `LookAndFeel` classes in order to hide details about default L&F selection within the `LookAndFeel` class.  See the second commit for details.
* I changed the version coordinate of the Substance dependency.  I was originally going to use this coordinate in #4001, but the Jitpack Ant builds weren't working at that time.  They've since been fixed, and this coordinate is now valid.

## Manual Testing Performed

* I applied each Substance L&F from the engine settings to ensure there weren't any copy/paste errors in the class names.
* Verified the default L&F was selected correctly when the client setting was not set (Linux only).
* Verified I could select a Substance L&F from the portable build to ensure the dependencies were packaged correctly.

## Additional Review Notes

There is a debatable change in this PR.  Namely, I chose to remove the compile-time dependency on the Substance library.  This then requires us to hard-code the Substance L&F class names instead of referring to the `Class` instances directly.

I didn't do this simply to avoid injecting the dependency from `game-headed` to `game-core`.  In fact, it turns out getting this dependency into `game-core` isn't that big of a deal and could be done quite easily.

My primary motivation for using class names instead of the classes themselves is to avoid loading over 25 L&F classes that will never be used.  I noticed that Swing follows this approach in that the `LookAndFeelInfo` type never loads a L&F class until it is needed.  I'm _assuming_ that loading a L&F class may cause some additional heavyweight resources to be loaded, otherwise Swing wouldn't go to the trouble of dealing with L&F class names almost exclusively.

Anyway, if there's a strong team preference to keep the compile-time dependency on the Substance L&F classes, as I said above, it wouldn't require much work to add it back.  The only benefit I see from our perspective in doing this is that, when we upgrade Substance, if a L&F class is ever removed or renamed, we'll detect it during the build rather than when a user reports an error at runtime.
